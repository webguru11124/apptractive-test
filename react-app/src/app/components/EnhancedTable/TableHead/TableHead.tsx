import React from 'react';
import { Checkbox, TableCell, TableRow } from '@mui/material';
import TableHead from '@mui/material/TableHead';
type HeaderCell = {
  label: string;
  numeric: 'right' | 'left' | 'center';
};
interface EnhancedTableProps {
  numSelected: number;
  onSelectAllClick: (event: React.ChangeEvent<HTMLInputElement>) => void;
  rowCount: number;
  headerCells: HeaderCell[];
}

export function EnhancedTableHead(props: EnhancedTableProps) {
  const { onSelectAllClick, numSelected, rowCount } = props;
  return (
    <TableHead>
      <TableRow>
        <TableCell padding="checkbox">
          <Checkbox
            color="primary"
            indeterminate={numSelected > 0 && numSelected < rowCount}
            checked={rowCount > 0 && numSelected === rowCount}
            onChange={onSelectAllClick}
            inputProps={{
              'aria-label': 'select all desserts',
            }}
          />
        </TableCell>
        {props.headerCells.map((headCell, index) => (
          <TableCell key={index} align={headCell.numeric ? 'right' : 'left'}>
            {headCell.label}
          </TableCell>
        ))}
      </TableRow>
    </TableHead>
  );
}
