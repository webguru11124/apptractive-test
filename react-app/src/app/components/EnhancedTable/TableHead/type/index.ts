export type Order = 'asc' | 'desc';
export type OrderBy = { id: string; order: Order };
