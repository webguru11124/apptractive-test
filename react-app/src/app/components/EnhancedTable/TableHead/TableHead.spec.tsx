import React from 'react';
import { render, screen } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import { MemoryRouter } from 'react-router-dom';

import { EnhancedTableHead, EnhancedTableProps } from './TableHead';
import { Order } from './type';

if (typeof Object.hasOwn !== 'function') {
  Object.hasOwn = function (object: any, property: any) {
    return Object.prototype.hasOwnProperty.call(object, property);
  };
}

// Clean up after each test
afterEach(() => {
  vi.restoreAllMocks();
});

const renderComponent = (props: EnhancedTableProps) =>
  render(
    <MemoryRouter>
      <table>
        <EnhancedTableHead {...props} />
      </table>
    </MemoryRouter>
  );

describe('EnhancedTableHead Component', () => {
  const defaultProps: EnhancedTableProps = {
    onSelectAllClick: vi.fn(),
    numSelected: 0,
    rowCount: 2,
    headerCells: [
      {
        id: 'header1',
        label: 'Header 1',
        numeric: 'left',
      },
      {
        id: 'header2',
        label: 'Header 2',
        numeric: 'center',
      },
      {
        id: 'header3',
        label: 'Header 3',
        numeric: 'right',
      },
    ],
    orderBy: 'header2',
    order: 'asc' as Order,
    onChangeOrder: vi.fn(),
  };

  it('renders the correct number of headers', async () => {
    renderComponent(defaultProps);
    const headers = screen.getAllByRole('columnheader');
    expect(headers).toHaveLength(defaultProps.headerCells.length + 1); // +1 for the checkbox column
  });

  it('checkbox is not checked if no items are selected', () => {
    renderComponent({ ...defaultProps, numSelected: 0, rowCount: 5 });
    const checkbox = screen.getByRole('checkbox') as HTMLInputElement;
    expect(checkbox.checked).toBe(false);
  });

  it('checkbox is checked if all items are selected', () => {
    renderComponent({ ...defaultProps, numSelected: 5, rowCount: 5 });
    const checkbox = screen.getByRole('checkbox') as HTMLInputElement;
    expect(checkbox.checked).toBe(true);
  });

  it('calls onSelectAllClick when checkbox is clicked', () => {
    const handleSelectAllClick = vi.fn();
    renderComponent({
      ...defaultProps,
      onSelectAllClick: handleSelectAllClick,
      numSelected: 0,
      rowCount: 5,
    });
    const checkbox = screen.getByRole('checkbox');
    checkbox.click();
    expect(handleSelectAllClick).toHaveBeenCalled();
  });
});
