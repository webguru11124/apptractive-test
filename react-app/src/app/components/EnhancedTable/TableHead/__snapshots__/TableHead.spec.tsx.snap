// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EnhancedTableHead Component > renders the correct number of headers 1`] = `
<thead
  class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
  data-testid="table-head"
>
  <tr
    class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
  >
    <th
      class="MuiTableCell-root MuiTableCell-head MuiTableCell-paddingCheckbox MuiTableCell-sizeMedium css-hbtnrh-MuiTableCell-root"
      scope="col"
    >
      <span
        class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          aria-label="select all desserts"
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
    </th>
    <th
      class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
      scope="col"
    >
      <span
        class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
        role="button"
        tabindex="0"
      >
        Header 1
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
          data-testid="ArrowDownwardIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
          />
        </svg>
      </span>
    </th>
    <th
      aria-sort="ascending"
      class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
      scope="col"
    >
      <span
        class="MuiButtonBase-root MuiTableSortLabel-root Mui-active css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
        role="button"
        tabindex="0"
      >
        Header 2
        <span
          class="MuiBox-root css-trfuzi"
        >
          sorted ascending
        </span>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
          data-testid="ArrowDownwardIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
          />
        </svg>
      </span>
    </th>
    <th
      class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
      scope="col"
    >
      <span
        class="MuiButtonBase-root MuiTableSortLabel-root css-1qgma8u-MuiButtonBase-root-MuiTableSortLabel-root"
        role="button"
        tabindex="0"
      >
        Header 3
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc css-1vweko9-MuiSvgIcon-root-MuiTableSortLabel-icon"
          data-testid="ArrowDownwardIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
          />
        </svg>
      </span>
    </th>
  </tr>
</thead>
`;
